2005-03-03 Enrico Scholz <enrico.scholz@informatik.tu-chemnitz.de>
	* fixed various places where signed and unsigned variables are
	  compared; most if them cause compiler warnings also, but
	  other might cause buffer overflows also

diff -ur hunt-1.5badcmp/arphijack.c hunt-1.5/arphijack.c
--- hunt-1.5badcmp/arphijack.c	2005-03-03 14:21:00.000000000 +0100
+++ hunt-1.5/arphijack.c	2005-03-03 14:27:29.000000000 +0100
@@ -57,7 +57,7 @@
 
 	if (wtd->input_mode == INPUT_MODE_RAW)
 		tty_raw(0, 1, 0);
-	while ((nr = read(0, buf, sizeof(buf)))) {
+	while ((nr = read(0, buf, sizeof(buf)))>0) {
 		if (buf[0] == 29)	/* ^] */
 			break;
 		if (wtd->input_mode == INPUT_MODE_LINEECHO || 
@@ -67,7 +67,7 @@
 				break;
 			
 			if (wtd->input_mode == INPUT_MODE_LINEECHOR && 
-			    nr < sizeof(buf) && buf[nr - 1] == '\n') {
+			    (size_t)nr < sizeof(buf) && buf[nr - 1] == '\n') {
 				buf[nr - 1] = '\r';
 				buf[nr++] = '\n';
 			}
diff -ur hunt-1.5badcmp/arpspoof.c hunt-1.5/arpspoof.c
--- hunt-1.5badcmp/arpspoof.c	2005-03-03 14:21:00.000000000 +0100
+++ hunt-1.5/arpspoof.c	2005-03-03 14:31:43.000000000 +0100
@@ -970,7 +970,7 @@
 	char buf[BUFSIZE];
 	int refresh, can_forward;
 	
-	if ((src_ip = menu_choose_hostname("host to spoof", NULL)) == -1)
+	if ((src_ip = menu_choose_hostname("host to spoof", NULL)) == (unsigned int)(-1))
 		return;
 	sprintf_eth_mac(buf, suggest_mac());
 	if (menu_choose_mac("fake mac", src_fake_mac, buf) < 0)
@@ -980,7 +980,7 @@
 			return;
 	} else
 		can_forward = 1;
-	if ((dst_ip = menu_choose_hostname("target - where to insert the spoof", NULL)) == -1)
+	if ((dst_ip = menu_choose_hostname("target - where to insert the spoof", NULL)) == (unsigned int)(-1))
 		return;
 	if ((refresh = menu_choose_unr("refresh interval sec", 0, 100000, 0)) < 0)
 		return;
@@ -1024,7 +1024,7 @@
 	char buf[BUFSIZE];
 	int refresh, can_forward;
 	
-	if ((src_ip = menu_choose_hostname("host to spoof", NULL)) == -1)
+	if ((src_ip = menu_choose_hostname("host to spoof", NULL)) == (unsigned int)(-1))
 		return;
 	sprintf_eth_mac(buf, suggest_mac());
 	if (menu_choose_mac("fake mac", src_fake_mac, buf) < 0)
@@ -1034,9 +1034,9 @@
 			return;
 	} else
 		can_forward = 1;
-	if ((dst_start_ip = menu_choose_hostname("start target where to insert the spoof", NULL)) == -1)
+	if ((dst_start_ip = menu_choose_hostname("start target where to insert the spoof", NULL)) == (unsigned int)(-1))
 		return;
-	if ((dst_end_ip = menu_choose_hostname("end target where to insert the spoof", NULL)) == -1)
+	if ((dst_end_ip = menu_choose_hostname("end target where to insert the spoof", NULL)) == (unsigned int)(-1))
 		return;
 	if ((refresh = menu_choose_unr("refresh interval sec", 0, 100000, 0)) < 0)
 		return;
@@ -1120,7 +1120,7 @@
 	struct mac_info *mi_src, *mi_dst;
 	int refresh, src_can_forward, dst_can_forward;
 	
-	if ((src_ip = menu_choose_hostname("src/dst host1 to arp spoof", NULL)) == -1)
+	if ((src_ip = menu_choose_hostname("src/dst host1 to arp spoof", NULL)) == (unsigned int)(-1))
 		return;
 	sprintf_eth_mac(buf, suggest_mac());
 	if (menu_choose_mac("host1 fake mac", src_fake_mac, buf) < 0)
@@ -1130,7 +1130,7 @@
 			return;
 	} else
 		src_can_forward = 1;
-	if ((dst_ip = menu_choose_hostname("src/dst host2 to arp spoof", NULL)) == -1)
+	if ((dst_ip = menu_choose_hostname("src/dst host2 to arp spoof", NULL)) == (unsigned int)(-1))
 		return;
 	sprintf_eth_mac(buf, suggest_mac());
 	if (menu_choose_mac("host2 fake mac", dst_fake_mac, buf) < 0)
@@ -1183,9 +1183,9 @@
 	i = menu_choose_unr("item nr. with src/dst or [cr]", 0,
 			    arp_spoof_count() - 1, -1);
 	if (i < 0) {
-		if ((ip1 = menu_choose_hostname("src/dst host1 to remove", NULL)) == -1)
+		if ((ip1 = menu_choose_hostname("src/dst host1 to remove", NULL)) == (unsigned int)(-1))
 			return;
-		if ((ip2 = menu_choose_hostname("src/dst host2 to remove", NULL)) == -1)
+		if ((ip2 = menu_choose_hostname("src/dst host2 to remove", NULL)) == (unsigned int)(-1))
 			return;
 	} else {
 		asi = list_at(&l_arp_spoof, i);
@@ -1257,17 +1257,17 @@
 	if (range_test)
 		dst_addr = (unsigned int) -1;
 	else
-		if ((dst_addr = menu_choose_hostname("host to test", NULL)) == -1)
+		if ((dst_addr = menu_choose_hostname("host to test", NULL)) == (unsigned int)(-1))
 			return;
 	for (i = 0; i < asr->asi_count; i++) {
-		if (dst_addr == -1 && asr->asi[i]->dst_mac_valid)
+	  if (dst_addr == (unsigned int)(-1) && asr->asi[i]->dst_mac_valid)
 			do_test_or_refresh(asr->asi[i]);
 		else if (asr->asi[i]->dst_addr == dst_addr) {
 			do_test_or_refresh(asr->asi[i]);
 			break;
 		}
 	}
-	if (dst_addr != -1 && i >= asr->asi_count)
+	if (dst_addr != (unsigned int)(-1) && i >= asr->asi_count)
 		printf("host not found in range database\n");
 }
 
diff -ur hunt-1.5badcmp/hostup.c hunt-1.5/hostup.c
--- hunt-1.5badcmp/hostup.c	2005-03-03 14:21:00.000000000 +0100
+++ hunt-1.5/hostup.c	2005-03-03 14:24:44.000000000 +0100
@@ -229,9 +229,9 @@
 	char buf_mac[BUFSIZE];
 	unsigned char fake_mac[ETH_ALEN];
 	
-	if ((start_ip = menu_choose_hostname("start ip addr", host_lookup(start_ip_def, HL_MODE_NR))) == -1)
+	if ((start_ip = menu_choose_hostname("start ip addr", host_lookup(start_ip_def, HL_MODE_NR))) == (unsigned int)(-1))
 		return;
-	if ((end_ip = menu_choose_hostname("end ip addr", host_lookup(end_ip_def, HL_MODE_NR))) == -1)
+	if ((end_ip = menu_choose_hostname("end ip addr", host_lookup(end_ip_def, HL_MODE_NR))) == (unsigned int)(-1))
 		return;
 	if ((len = ntohl(end_ip) - ntohl(start_ip) + 1) < 0) {
 		printf("bad addresses\n");
diff -ur hunt-1.5badcmp/hunt.h hunt-1.5/hunt.h
--- hunt-1.5badcmp/hunt.h	2005-03-03 14:21:00.000000000 +0100
+++ hunt-1.5/hunt.h	2005-03-03 14:25:04.000000000 +0100
@@ -566,7 +566,7 @@
  * macdisc
  */
 void mac_discover(unsigned int ip, int count);
-void mac_discover_range(unsigned int start_ip, unsigned int end_ip, int count);
+void mac_discover_range(unsigned int start_ip, unsigned int end_ip, unsigned int count);
 void mac_disc_menu(void);
 void print_mac_daemon();
 
diff -ur hunt-1.5badcmp/macdisc.c hunt-1.5/macdisc.c
--- hunt-1.5badcmp/macdisc.c	1999-02-21 20:21:37.000000000 +0100
+++ hunt-1.5/macdisc.c	2005-03-03 14:24:44.000000000 +0100
@@ -14,7 +14,7 @@
 #include <assert.h>
 #include <errno.h>
 
-void mac_discover_range(unsigned int start_ip, unsigned int end_ip, int count)
+void mac_discover_range(unsigned int start_ip, unsigned int end_ip, unsigned int count)
 {
 	unsigned int addr, j;
 	struct timespec ts;
@@ -184,9 +184,9 @@
 	struct mac_disc_info *mdi;
 	unsigned int start_ip, end_ip;
 	
-	if ((start_ip = menu_choose_hostname("start ip addr", NULL)) == -1)
+	if ((start_ip = menu_choose_hostname("start ip addr", NULL)) == (unsigned int)(-1))
 		return;
-	if ((end_ip = menu_choose_hostname("end ip addr", NULL)) == -1)
+	if ((end_ip = menu_choose_hostname("end ip addr", NULL)) == (unsigned int)(-1))
 		return;
 	mdi = malloc(sizeof(struct mac_disc_info));
 	assert(mdi);
@@ -207,9 +207,9 @@
 	if (!(mdi = list_at(&l_mdi, nr)))
 		return;
 
-	if ((start_ip = menu_choose_hostname("start ip addr", host_lookup(mdi->start_addr, hl_mode))) == -1)
+	if ((start_ip = menu_choose_hostname("start ip addr", host_lookup(mdi->start_addr, hl_mode))) == (unsigned int)(-1))
 		return;
-	if ((end_ip = menu_choose_hostname("end ip addr", host_lookup(mdi->end_addr, hl_mode))) == -1)
+	if ((end_ip = menu_choose_hostname("end ip addr", host_lookup(mdi->end_addr, hl_mode))) == (unsigned int)(-1))
 		return;
 	mdi->start_addr = start_ip;
 	mdi->end_addr = end_ip;
diff -ur hunt-1.5badcmp/main.c hunt-1.5/main.c
--- hunt-1.5badcmp/main.c	2000-05-26 16:06:39.000000000 +0200
+++ hunt-1.5/main.c	2005-03-03 14:31:19.000000000 +0100
@@ -145,7 +145,7 @@
 		 * packet from source
 		 */
 		if (!dst_packet && p->p_data_len && !same_chars) {
-			pbuf_len = p->p_data_len < sizeof(pbuf) ? p->p_data_len : sizeof(pbuf);
+			pbuf_len = ((size_t)(p->p_data_len) < sizeof(pbuf)) ? (size_t)p->p_data_len : sizeof(pbuf);
 			memcpy(pbuf, p->p_data, pbuf_len);
 		} else
 			pbuf_len = 0;
diff -ur hunt-1.5badcmp/menu.c hunt-1.5/menu.c
--- hunt-1.5badcmp/menu.c	2005-03-03 14:21:00.000000000 +0100
+++ hunt-1.5/menu.c	2005-03-03 14:24:44.000000000 +0100
@@ -264,7 +264,7 @@
 	while (1) {
 		if (menu_prompt(label, buf, sizeof(buf), dfl) < 0)
 			return -1;
-		if ((ip = parse_hostname(buf)) != -1)
+		if ((ip = parse_hostname(buf)) != (unsigned int)(-1))
 			break;
 	}
 	return ip;
@@ -302,7 +302,7 @@
 		
 		if (!(host_name = strtok(buf, " /\t\n")))
 			continue;
-		if ((ip = parse_hostname(host_name)) == -1)
+		if ((ip = parse_hostname(host_name)) == (unsigned int)(-1))
 			continue;
 		if (with_mask) {
 			if (!(mask_str = strtok(NULL, " \t\n")))
diff -ur hunt-1.5badcmp/resolv.c hunt-1.5/resolv.c
--- hunt-1.5badcmp/resolv.c	1999-02-14 13:59:13.000000000 +0100
+++ hunt-1.5/resolv.c	2005-03-03 14:30:37.000000000 +0100
@@ -411,10 +411,10 @@
 char *host_lookup(unsigned int in, int use_mode)
 {
  	static char hostname_buf[BUFSIZE] = {0};
-	static int hostname_idx = 0;
+	static unsigned int hostname_idx = 0;
 	struct in_addr addr;
 	char *name, *retval;
-	int len;
+	size_t len;
 	struct hostent *host_ent;
 	struct resolv_item *r;
 	
@@ -495,10 +495,10 @@
 char *port_lookup(unsigned short serv, int use_mode)
 {
  	static char servname_buf[BUFSIZE] = {0};
-	static int servname_idx = 0;
+	static unsigned int servname_idx = 0;
 	char name_buf[64];
 	char *name, *retval;
-	int len;
+	size_t len;
 	struct servent *serv_ent;
 
 	if (serv == 0)
diff -ur hunt-1.5badcmp/synchijack.c hunt-1.5/synchijack.c
--- hunt-1.5badcmp/synchijack.c	1999-02-21 21:04:36.000000000 +0100
+++ hunt-1.5/synchijack.c	2005-03-03 14:30:37.000000000 +0100
@@ -197,7 +197,7 @@
 {
 	struct tcp_spec ts;
 	char buf[1400];
-	int len;
+	unsigned int len;
 
 	/*
 	 * well - after sending something we get usualy ack storm
@@ -257,7 +257,7 @@
 			memset(buf, ' ', len);
 #else
 		len = need_write;
-		if (len > sizeof(buf)) {
+		if ((size_t)len > sizeof(buf)) {
 			len = sizeof(buf);
 			memset(buf, ' ', len);
 		} else {
@@ -377,7 +377,7 @@
 			print_data_packet(p, p->p_data_len, 0, 0);
 		if (last_read_ack == p->p_hdr.p_tcph->ack_seq && p->p_data_len) {
 			len = p->p_data_len;
-			if (len > sizeof(buf))
+			if ((size_t)len > sizeof(buf))
 				len = sizeof(buf);
 			memcpy(buf, p->p_data, len);
 			w_data = buf;
